daemon/containerd/image_pull.go

```diff
@@ -42,9 +44,25 @@ func (i *ImageService) PullImage(ctx context.Context, image, tagOrDigest string,
 		}
 	}
 
-	resolver := newResolverFromAuthConfig(authConfig)
+	resolver, _ := i.newResolverFromAuthConfig(authConfig)
 	opts = append(opts, containerd.WithResolver(resolver))
 
+	jobs := newJobs()
+	h := images.HandlerFunc(func(ctx context.Context, desc specs.Descriptor) ([]specs.Descriptor, error) {
+		if desc.MediaType != images.MediaTypeDockerSchema1Manifest {
+			jobs.Add(desc)
+		}
+		return nil, nil
+	})
```

Comment:
This `jobs` value and its associated handler implementation smell like an implementation detail of `pullProgress`. Could those details be abstracted away inside `progress.go`?

Spitballing:
```go
progress := pullProgress(ctx, out, i.client.ContentStore())
defer progress.Stop()
opts = append(opts, containerd.WithImageHandler(progress))
```

Reply:
I'm not sure we can do that easily and in a way that would make the code more readable, once again, looking at the [push code](https://github.com/rumpl/moby/blob/c8d/daemon/containerd/image_push.go#L70-L79) the jobs we want to add are quite different.

Reply:
Ah, now I get it. You're totally right; the code is more readable the way you have it.